webapp-django/crashstats/cron/management/commands/cron.py

```diff
@@ -0,0 +1,464 @@
+# crontabber handles the jobs that haven't been converted, yet.
+#
+# NOTE(willkg): You can't have the same job in here twice with two different
+# frequencies. If we ever need to do that, we'll need to stop using "cmd" as
+# the key in the db.
+3
+# cmd, frequency, time, backfill
+JOBS = [
+    {
+        # Test cron job
+        'cmd': 'crontest',
+        'frequency': '1d',
+    },
+]
+
+# Map of cmd -> job_spec
+JOBS_MAP = dict([(job_spec['cmd'], job_spec) for job_spec in JOBS])
+
+
+# The maximum time we let a job go for before we declare it a zombie (in seconds)
+MAX_ONGOING = 60 * 60 * 2
+
+
+# Error retry time (in seconds)
+ERROR_RETRY_TIME = 60
+
+
+class FrequencyDefinitionError(Exception):
+    pass
+
+
+class JobNotFoundError(Exception):
+    pass
+
+
+class TimeDefinitionError(Exception):
+    pass
+
+
+class OngoingJobError(Exception):
+    """Error raised when a job is currently running."""
+
+    pass
+
+
+FREQUENCY_RE = re.compile(r'^(\d+)([^\d])$')
+
+
+def convert_frequency(value):
+    """return the number of seconds that a certain frequency string represents.
+    For example: `1d` means 1 day which means 60 * 60 * 24 seconds.
+    The recognized formats are:
+        10d  : 10 days
+        3m   : 3 minutes
+        12h  : 12 hours
+    """
+    match = FREQUENCY_RE.match(value)
+    if not match:
+        raise FrequencyDefinitionError(value)
+
+    number = int(match.group(1))
+    unit = match.group(2)
+
+    if unit == 'h':
+        number *= 60 * 60
+    elif unit == 'm':
+        number *= 60
+    elif unit == 'd':
+        number *= 60 * 60 * 24
+    elif unit:
+        raise FrequencyDefinitionError(value)
+    return number
+
+
+VALID_TIME_RE = re.compile(r'^\d\d:\d\d$')
+
+
+def check_time(value):
+    """Validate it's an appropriate time in HH:MM form."""
+    if not VALID_TIME_RE.match(value):
+        raise TimeDefinitionError("Invalid definition of time %r" % value)
+
+    hh, mm = value.split(':')
+    if int(hh) > 23 or int(mm) > 59:
+        raise TimeDefinitionError("Invalid definition of time %r" % value)
+
+
+def get_matching_job_specs(cmds):
+    """Return list of matching job specs for this cmd.
+
+    :arg cmds: "all" returns list of all job_specs, cmds return single job_spec,
+        list of cmds returns list of job_specs
+
+    """
+    if cmds == ['all']:
+        return JOBS
+    if isinstance(cmds, str):
+        return JOBS_MAP.get(cmds)
+    return [JOBS_MAP[cmd] for cmd in cmds if cmd in JOBS_MAP]
+
+
+def time_to_run(job_spec, job):
+    """Determine whether it's time for this cmd to run."""
+    now = timezone.now()
+    time_ = job_spec.get('time')
+
+    if job.next_run is None:
+        if time_:
+            # Only run if this hour and minute is < now
+            h, m = [int(x) for x in time_.split(':')]
+            if now.hour > h:
+                return True
+            elif now.hour == h and now.minute >= m:
+                return True
```

Comment:
Could probably redo this as:

```
if (now.hour, now.minute) >= (h, m):
```

That'll kick off at TIME or later.

Reply:
Also, interestingly `check_time` isn't used anywhere. We might want to nix that and have it parse out the time. So then it'd return None for no time, (h, m) for a time, and raise TimeDefinitionError if it parses poorly.